### ISSUE TITLE
Cannot get data with server-only modules using data loaders

### ISSUE THREAD

**Issue by tundera** [jacob-ebey/mwap#7]
I'm having some trouble with the Data Loaders concept. It seems pretty straightforward, but when I try to use a module that's intended for use only on the server, I get an error letting me know that module is not intended to be used in the browser. So I'm wondering where/when the data-fetching functions I define in the data loaders are intended to be executed. Is there a way to use server-side modules for fetching data on the server and passing it along to a page (kind of like `getServerSideProps` in Next)?

Here's an example with Prisma (`db` is just a prisma client instance):

```ts
import type { Loader, LoaderConfigFunc } from 'mwap'
import db, { Team } from '../db'

export type TeamsPageData = {
  teams: Team[]
  ttl: number
}

const loader: Loader<TeamsPageData> = async () => {
  const teams = await db.team.findMany()

  return {
    teams,
    ttl: 604800,
  }
}

const config: LoaderConfigFunc<TeamsPageData> = ({ data }) => ({
  headers: {
    'cache-control': `public, max-age=${data.ttl}`,
  },
})

export default { loader, config }
```

In the browser navigating to the `teams` page where this data loader is used, I get the following error:
```sh
Error: PrismaClient is unable to be run in the browser.
In case this error is unexpected for you, please report it in https://github.com/prisma/prisma/issues
    at new PrismaClient (/Users/philip.johnston/Tundera/apps/mwap-app/dist/server/webpack:/mwap-app/node_modules/.prisma/client/index-browser.js:117:11)
    at Module../db/index.ts (/Users/philip.johnston/Tundera/apps/mwap-app/dist/server/webpack:/mwap-app/db/index.ts:9:24)
    at __webpack_require__ (/Users/philip.johnston/Tundera/apps/mwap-app/dist/server/webpack:/mwap-app/webpack/bootstrap:19:1)
    at Module../loaders/teams.ts (/Users/philip.johnston/Tundera/apps/mwap-app/dist/server/main.js:268:61)
    at __webpack_require__ (/Users/philip.johnston/Tundera/apps/mwap-app/dist/server/webpack:/mwap-app/webpack/bootstrap:19:1)
    at Module../loaders/index.ts (/Users/philip.johnston/Tundera/apps/mwap-app/dist/server/main.js:243:64)
    at __webpack_require__ (/Users/philip.johnston/Tundera/apps/mwap-app/dist/server/webpack:/mwap-app/webpack/bootstrap:19:1)
    at /Users/philip.johnston/Tundera/apps/mwap-app/dist/server/webpack:/mwap-app/node_modules/@mwap/server/dist/server.modern.js:72:18
    at Generator.next (<anonymous>)
    at /Users/philip.johnston/Tundera/apps/mwap-app/dist/server/webpack:/mwap-app/node_modules/@mwap/server/dist/server.modern.js:20:61
```

**Comment by jacob-ebey** [comment]
Can you put up a simple repo case? Potentially make your db import:

```js
if (typeof window !== "undefined") {
  throw new Error("Something is F'ed");
}
export type Team = string;
export default {
  team: { findmany: () => [] }
}
```

Loaders should only run / be imported / bundled with server side code. If you want to dig, clone this repo down and create the repo in the `examples` folder.

I have stuff to attend to today but will try to look at this ASAP.

**Comment by jacob-ebey** [comment]
Ok, I've published an update in v0.0.9 of the CLI that maps mainFields appropriately. There does however seem to be an issue with prisma in general in webpack5 based projects: https://github.com/prisma/prisma/issues/6925

I was able to get it imported and an error from prisma about the query engine binary missing (I don't use prisma so IDK how to set it up).

You can create a `mwap.config.js` in your project root that looks like this to get to the point I got to:

```js
module.exports = {
  webpack(config, { isServer }) {
    if (isServer) {
      config.externals = config.externals || {};
      config.externals.encoding = "commonjs @prisma/client";
      config.externals.encoding = "commonjs encoding";
      config.externals._http_common = "commonjs _http_common";
    }

    return config;
  },
};
```

Can we call this closed if you can get to the same failure point I did? If there are other issues not related to the mainFields fix that are not prisma specific let's open another issue :)

Good catch and call out on the mainFields issue.

**Comment by tundera** [comment]
@jacob-ebey Definitely some stuff going on with Prisma it seems. Thanks for looking into it. Going with a different database client in the loaders seemed to work well. I think this issue is fine to close out 😄
